src/dataflow/src/source/postgres.rs

```diff
@@ -0,0 +1,319 @@
+
+        (rel_id, col_types)
+    }
+
+    async fn produce_snapshot(
+        &mut self,
+        client: &mut ReplicationClient,
+        col_types: &[PgColumn],
+        timestamper: &Timestamper,
+        co: &mut Co<StreamItem>,
+    ) -> (String, PgLsn) {
+        // TODO: find better naming scheme. Postgres doesn't like dashes
+        let slot_name = Uuid::new_v4().to_string().replace('-', "");
+
+        // Start a transaction and immediatelly create a replication slot with the USE SNAPSHOT
+        // directive. This makes the starting point of the slot and the snapshot of the transaction
+        // identical.
+        //
+        // TODO: The slot should be permanent and destroyed on Drop. This way we can support
+        // reconnects
+        client
+            .simple_query("BEGIN READ ONLY ISOLATION LEVEL REPEATABLE READ;")
+            .await
+            .unwrap();
+        let slot = client
+            .create_logical_replication_slot(
+                &slot_name,
+                true,
+                &self.plugin,
+                Some(SnapshotMode::UseSnapshot),
+            )
+            .await
+            .unwrap();
+
+        let cur_lsn = slot.consistent_point();
+
+        // Initial data copy
+        let copy_query = format!(
+            r#"COPY "{}"."{}" TO STDOUT WITH (FORMAT binary);"#,
+            &self.connector.namespace, &self.connector.table
+        );
+
+        let stream = client.copy_out(&copy_query).await.unwrap();
+
+        let scalar_types = col_types
+            .iter()
+            .map(|c| c.scalar_type.clone())
+            .collect_vec();
+        let rows = BinaryCopyOutStream::new(stream, &scalar_types);
+        pin_mut!(rows);
+
+        let snapshot_ts = timestamper.lease().await;
+
+        // TODO: handle Err case
+        while let Some(Ok(row)) = rows.next().await {
+            self.packer.clear();
+            for (i, ty) in col_types.iter().enumerate() {
+                let datum: Datum = match ty.scalar_type {
+                    PgType::BOOL => row.get::<Option<bool>>(i).into(),
+                    PgType::INT4 => row.get::<Option<i32>>(i).into(),
+                    PgType::INT8 => row.get::<Option<i64>>(i).into(),
+                    PgType::FLOAT4 => row.get::<Option<f32>>(i).into(),
+                    PgType::FLOAT8 => row.get::<Option<f64>>(i).into(),
+                    PgType::DATE => row.get::<Option<NaiveDate>>(i).into(),
+                    PgType::TIME => row.get::<Option<NaiveTime>>(i).into(),
+                    PgType::TIMESTAMP => row.get::<Option<NaiveDateTime>>(i).into(),
+                    PgType::TEXT => row.get::<Option<&str>>(i).into(),
+                    PgType::UUID => row.get::<Option<Uuid>>(i).into(),
+                    ref other => todo!("Unsupported data type {:?}", other),
+                };
+
+                if !ty.nullable && datum.is_null() {
+                    panic!("null value in non nullable column");
```

Comment:
I know its WIP but we probably don't want to panic here right?

Reply:
Yes, I need to wire up the error stream for a bunch of cases like this one or any of the 10s of `unwrap()` calls :)